### ISSUE TITLE
Board Review: <client library name>

### ISSUE THREAD

**Issue by jsquire** [Azure/azure-sdk#1367]
## The Basics

* Service team responsible for the client library: Event Hubs
* Link to documentation describing the service: https://azure.microsoft.com/en-us/services/event-hubs/
* Contact email (if service team, provide PM and Dev Lead): SeongJoon Kwak, Javier Fernandez , Shubha Vijayasarathy

## About this client library

* Name of the client library: azure-messaging-eventhubs
* Languages for this review: Python, Java, .NET, JS
* Link to the service REST APIs: **_N/A_**

### .NET

* [API View](https://apiview.dev/Assemblies/Review/8593bbb9252f4f1aaf1a616715c8d034?diffRevisionId=d20d5e77123a4e71bcbff97b5dbebcb1)
* [Design Proposal (includes samples)](https://gist.github.com/jsquire/f0948c9a5ffe2741b03d312dcf8d899e)

### Java

* Upload JAR to [apiview](https://apiview.azurewebsites.net).  Link:
* Link to samples for champion scenarios:

### Python

* Upload the api as a Draft PR.  Link to PR:
* Link to samples for champion scenarios:

### TypeScript

* Upload output of api-extractor as a Draft PR.  Link to PR:
* Link to samples for champion scenarios:

### Connecting on a restricted network

A local hospital employs a strict set of network policies to ensure the safety of their network and the security of their patient information. As part of this strategy, all traffic is routed through a security proxy which performs a threat analysis and logging of connections.

In order to allow for trusted internal applications to embrace cloud services, a special bank of IP addresses has been reserved which passes through a secure DMZ path rather than the security proxy. In order for trusted connections to work appropriately, they must resolve to one of the reserved IP addresses using a CNAME configured in the local DNS service.

Because of these restrictions, applications are unable to perform direct connections to the Azure Event Hubs service using the standard endpoint address and require the ability to specify a custom host name to ensure they route through the proper intermediary for the connection to be made.

**Comment by jsquire** [comment]
//cc: @KrzysztofCwalina

**Comment by KrzysztofCwalina** [comment]
This (one additional property on client options) looks good to me.

**Comment by bterlson** [comment]
Same here.

**Comment by johanste** [comment]
Is this only an issue for EventHubs?

**Comment by JonathanGiles** [comment]
Fine with me.

**Comment by jsquire** [comment]
> Is this only an issue for EventHubs?

This will also be an issue in Service Bus, due to the Messaging infrastructure for routing to a node in their cluster.  It should not, to my knowledge, impact other services.

**Comment by johanste** [comment]
Should it be considered a proxy setting?

**Comment by jsquire** [comment]
I don't believe that would work, unless I'm misunderstanding the question.  For the Messaging clients, a proxy can only be used when `WebSockets` are selected as the transport.  The custom endpoint needs to be applied for both `TCP` and `WebSocket`. 

For some languages, like C#, we're expecting a built-in framework type for the proxy, which is provided by the caller and for which there are no additional options passed.  The proxy as well as the other transport values are supplied as part of the `EventHubConnectionOptions`, which is where this setting was added, since it is also related to how connections are established.

**Comment by JeffreyRichter** [comment]
I spoke to Jesse and I'm OK with this.
